app/state/[state]/session/[session]/bill/[bill]/saveButton.tsx

```diff
@@ -0,0 +1,105 @@
+'use client'
+
+import { FC, useState, useEffect } from 'react';
+import SaveButtonProps from '@/data/saveButtonProps';
+
+const SaveButton: FC<SaveButtonProps> = ({ bill }) => {
+  const [isSaved, setIsSaved] = useState(false);
+  const [savedBillId, setSavedBillId] = useState<number | null>(null);
+  const [loading, setLoading] = useState(true);
+
+  const legiscanBillId = bill.bill_id || bill.legiscanBillId;
+
+  // Check if the bill is already saved when the component mounts
+  useEffect(() => {
+
+    const fetchIsSaved = async () => {
+      try {
+        const response = await fetch(`/api/checkSavedBill?legiscanBillId=${legiscanBillId}`);
+        const result = await response.json();
+        if (result.savedBill) {
+          setIsSaved(true);
+          setSavedBillId(result.savedBill.id); // Save the id for unsaving later
+        }
+      } catch (error) {
+        console.error('Error checking if bill is saved:', error);
+      } finally {
+        setLoading(false); // Ensure loading state is false regardless of success or error
+      }
+    };
+
+    fetchIsSaved();
+  }, [bill]);
+
+  const handleClick = async () => {
+    if (!bill) {
+      console.error('Bill object is undefined');
+      return;
+    }
+
+    try {
+      // Optimistically update the state before the async operation
+      setIsSaved(!isSaved);
+
+      if (isSaved && savedBillId) {
+        // If the bill is saved, should remove it
+        const response = await fetch('/api/unsaveBill', {
+          method: 'DELETE',
+          headers: {
+            'Content-Type': 'application/json',
+          },
+          body: JSON.stringify({ id: savedBillId }),
+        });
+
+        if (!response.ok) {
+          // If the request fails, revert the state
+          setIsSaved(true);
+          console.error('Failed to unsave bill');
+        } else {
+          setSavedBillId(null); // Clear the saved bill ID
+        }
+      } else {
+        // If the bill is not saved, save it
+        const response = await fetch('/api/saveBill', {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json',
+          },
+          body: JSON.stringify({
+            legiscanBillId: bill.bill_id,
```

Comment:
Bug here. Resaving a billl in the dashboard after unsaving creates a server side error: 

```
Error saving bill: PrismaClientValidationError: 
Invalid `prisma.savedBill.create()` invocation:

{
  data: {
    userId: 1,
    title: "House resolution relating to House Rules",
    description: "House resolution relating to House Rules",
    state: "VT",
    sessionTitle: "Unknown Session",
    sessionId: 0,
    changeHash: undefined,
+   legiscanBillId: Int
  }
}

Argument `legiscanBillId` is missing.
    at An (/Users/zomb/code/playground/legi/legibill/node_modules/@prisma/client/runtime/library.js:114:7526)
    at _n.handleRequestError (/Users/zomb/code/playground/legi/legibill/node_modules/@prisma/client/runtime/library.js:121:7392)
    at _n.handleAndLogRequestError (/Users/zomb/code/playground/legi/legibill/node_modules/@prisma/client/runtime/library.js:121:7057)
    at _n.request (/Users/zomb/code/playground/legi/legibill/node_modules/@prisma/client/runtime/library.js:121:6741)
    at async l (/Users/zomb/code/playground/legi/legibill/node_modules/@prisma/client/runtime/library.js:130:9355)
    at async POST (webpack-internal:///(rsc)/./app/api/saveBill/route.ts:26:27)
    at async /Users/zomb/code/playground/legi/legibill/node_modules/next/dist/compiled/next-server/app-route.runtime.dev.js:6:62609 {
  clientVersion: '5.18.0'
}
```

The `legiscanBillId` property should be set with `legiscanBillId` value defined at the top of this component. When the bill is unsaved, it is still shaped as the bill returned back from the db so it doesn't have a `bill_id`.

Reply:
This pr should resolve it. I had to ensure that it could receive both named data types from either legiscan or our db (ie bill_id or legiscanBillId, change_hash or changeHash).
I also found that clicking the button quickly in succession would save the same bill multiple times so I added an isProcessing state to prevent that.